specifyweb/frontend/js_src/lib/components/Attachments/RecordSetAttachment.tsx

```diff
@@ -0,0 +1,165 @@
+import React from 'react';
+
+import { useAsyncState } from '../../hooks/useAsyncState';
+import { useBooleanState } from '../../hooks/useBooleanState';
+import { useCachedState } from '../../hooks/useCachedState';
+import { attachmentsText } from '../../localization/attachments';
+import { commonText } from '../../localization/common';
+import { f } from '../../utils/functools';
+import type { RA } from '../../utils/types';
+import { filterArray } from '../../utils/types';
+import { Button } from '../Atoms/Button';
+import { serializeResource } from '../DataModel/helpers';
+import type { AnySchema, SerializedResource } from '../DataModel/helperTypes';
+import type { SpecifyResource } from '../DataModel/legacyTypes';
+import type {
+  Attachment,
+  CollectionObjectAttachment,
+} from '../DataModel/types';
+import { Dialog } from '../Molecules/Dialog';
+import { defaultAttachmentScale } from '.';
+import { AttachmentGallery } from './Gallery';
+
+const haltIncrementSize = 300;
+
+export function RecordSetAttachments<SCHEMA extends AnySchema>({
+  records,
+  onFetch: handleFetch,
+}: {
+  readonly records: RA<SpecifyResource<SCHEMA> | undefined>;
+  readonly onFetch:
+    | ((index: number) => Promise<RA<number | undefined> | void>)
+    | undefined;
+}): JSX.Element {
+  const fetchedCount = React.useRef<number>(0);
+
+  const [showAttachments, handleShowAttachments, handleHideAttachments] =
+    useBooleanState();
+
+  const attachmentsRef = React.useRef<{
+    attachments: RA<SerializedResource<Attachment>>;
+    related: RA<SpecifyResource<CollectionObjectAttachment>>;
+  }>();
+
+  const [attachments] = useAsyncState(
+    React.useCallback(async () => {
+      const relatedAttachmentRecords = await Promise.all(
+        records.map((record) =>
+          record
+            ?.rgetCollection(`${record.specifyModel.name}Attachments`)
+            .then(
+              ({ models }) =>
+                models as RA<SpecifyResource<CollectionObjectAttachment>>
+            )
+        )
+      );
+
+      const fetchCount = records.findIndex(
+        (record) => record?.populated !== true
+      );
+
+      fetchedCount.current = fetchCount === -1 ? records.length : fetchCount;
+
+      const attachements = await Promise.all(
+        filterArray(relatedAttachmentRecords.flat()).map(
+          async (collectionObjectAttachment) => ({
+            attachment: await collectionObjectAttachment
+              .rgetPromise('attachment')
+              .then((resource) => serializeResource(resource)),
+            related: collectionObjectAttachment,
+          })
+        )
+      );
+
+      const newAttachments = {
+        attachments: attachements.map(({ attachment }) => attachment) as RA<
+          SerializedResource<Attachment>
+        >,
+        related: attachements.map(({ related }) => related) as RA<
+          SpecifyResource<CollectionObjectAttachment>
+        >,
+      };
+
+      attachmentsRef.current = newAttachments;
```

Comment:
this can be simplified. just return new attachments like you did before, but right after the useEffect add this code:
```
if (typeof attachments === 'object') attachmentsRef.current = attachments;
```

Reply:
Then, you don't need to define `currentAttachments` and can just use `attachmentsRef.current`